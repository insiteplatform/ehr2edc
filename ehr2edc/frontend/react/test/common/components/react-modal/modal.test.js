import React from 'react';
import Modal, {ModalErrors} from "../../../../src/common/components/react-modal/modal";
import RequestError from "../../../../src/common/components/error/backend-request-error";

const title = "test";
const content = "Test";
const error = {
    response: {
        data: {
            "issues": [{
                "reference": null,
                "field": null,
                "message": "Something happened in the back-end"
            }]
        }
    }
};
const validationErrors = [
    "Validation failed.",
    "Another validation failed"
];

describe("Rendering the component", () => {
    test("Rendering the open state", () => {
        let open = true;
        const app = mount(<Modal open={open} title={title} onClose={() => open = false}><p>{content}</p></Modal>);
        expect(app.prop("open")).toEqual(true);

        expect(app).toMatchSnapshot();
        validateOpenModal(app, title, content);
    });
    test("Rendering the open state with an action in Progress", () => {
        let open = true;
        const app = mount(<Modal open={open} title={title} inProgress={true}
                                 onClose={() => open = false}><p>{content}</p></Modal>);
        expect(app.prop("open")).toEqual(true);

        expect(app).toMatchSnapshot();
        validateOpenModal(app, title, content);
    });
    test("Rendering the closed state", () => {
        const app = mount(<Modal open={false} title={title}><p>{content}</p></Modal>);
        expect(app.prop("open")).toEqual(false);

        expect(app).toMatchSnapshot();
        validateClosedModal(app);
    });
    test("A modal is in the close state by default", () => {
        const app = mount(<Modal title={title}><p>{content}</p></Modal>);
        expect(app.prop("open")).toEqual(undefined);

        expect(app).toMatchSnapshot();
        validateClosedModal(app);
    });
    test("A modal with a fixed width and height", () => {
        const app = mount(<Modal title={title} open={true} width="800px" height="350px"><p>{content}</p></Modal>);
        expect(app.prop("open")).toEqual(true);

        expect(app).toMatchSnapshot();
        validateOpenModal(app, title, content);
    });
    test("Reloading the component", () => {
        const app = mount(<Modal title={title} open={false}><p>{content}</p></Modal>);
        app.setProps({open: true});
        app.update();

        expect(app.prop("open")).toEqual(true);
        validateOpenModal(app, title, content);
    });
    test("Reloading the component without any data", () => {
        const app = mount(<Modal title={title} open={false}><p>{content}</p></Modal>);

        app.setProps({});

        expect(app.prop("open")).toEqual(false);
        validateClosedModal(app);
    });
});

describe("Actions", () => {
    test("Clicking the close icon puts the modal in the closed state", () => {
        let open = true;
        const app = shallow(<Modal open={open} title={title} onClose={() => open = false}><p>{content}</p></Modal>);

        app.find(".modal-close-icon").simulate('click', {});
        app.setProps({open: open});
        app.update();

        expect(app.prop("isOpen")).toEqual(open);
        validateClosedModal(app);
    });
    test("Clicking the cancel button puts the modal in the closed state", () => {
        let open = true;
        const app = shallow(<Modal open={open} title={title} onClose={() => open = false}><p>{content}</p></Modal>);

        app.find(".modal-cancel").simulate('click', {});
        app.setProps({open: open});
        app.update();

        expect(app.prop("isOpen")).toEqual(open);
        validateClosedModal(app);
    });
    test("Clicking the OK-button fires the onOkButtonClick event", () => {
        let onOkButtonClick = jest.fn();
        let app = shallow(<Modal open={true} title={title} onOkButtonClick={onOkButtonClick}><p>{content}</p></Modal>);

        app.find(".modal-ok").simulate('click', {});
        app.update();

        expect(onOkButtonClick.mock.calls.length).toEqual(1);
    });
});

describe("ModalErrors", () => {
    test("Renders validationErrors correctly", () => {
        const modalErrors = shallow(<ModalErrors validationErrors={validationErrors}/>, { disableLifecycleMethods: true });

        expect(modalErrors.find("li")).toHaveLength(2);
        expect(modalErrors.find("RequestError").prop("error")).toBeUndefined();
        expect(modalErrors).toMatchSnapshot();
    });
    test("Renders requestError correctly", () => {
        const modalErrors = shallow(<ModalErrors requestError={error} contact={"cont@ct"}/>, { disableLifecycleMethods: true });

        expect(modalErrors.find("li")).toHaveLength(0);
        expect(modalErrors.exists("RequestError")).toBe(true);
        expect(modalErrors.find("RequestError").prop("error")).toEqual(error);
        expect(modalErrors).toMatchSnapshot();
    });
    test("Renders combination correctly", () => {
        const modalErrors = shallow(<ModalErrors validationErrors={validationErrors}
                                                 requestError={error}
                                                 contact={"cont@ct"}/>, { disableLifecycleMethods: true });

        expect(modalErrors.find("li")).toHaveLength(2);
        expect(modalErrors.exists("RequestError")).toBe(true);
        expect(modalErrors.find("RequestError").prop("error")).toEqual(error);
        expect(modalErrors).toMatchSnapshot();
    })
});

function validateOpenModal(wrapper, title, content) {
    expect(wrapper.find('ModalPortal').children().length).toEqual(1);
    expect(wrapper.find('.modal-title').text()).toEqual(title);
    expect(wrapper.find('.modal-content').text()).toEqual(content);
    expect(wrapper.find('.modal-ok').text()).toEqual("OK");
    expect(wrapper.find('.modal-cancel').text()).toEqual("Cancel");
}

function validateClosedModal(wrapper) {
    expect(wrapper.find('ModalPortal').children().length).toEqual(0);
}