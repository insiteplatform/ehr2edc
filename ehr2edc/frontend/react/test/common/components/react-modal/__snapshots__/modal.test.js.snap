// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalErrors Renders combination correctly 1`] = `
<div
  className="alert-box alert margin-top"
>
  The form returned some errors
  <ul
    className="bullet"
  >
    <li
      key="0"
    >
      Validation failed.
    </li>
    <li
      key="1"
    >
      Another validation failed
    </li>
  </ul>
  <RequestError
    contact="cont@ct"
    error={
      Object {
        "response": Object {
          "data": Object {
            "issues": Array [
              Object {
                "field": null,
                "message": "Something happened in the back-end",
                "reference": null,
              },
            ],
          },
        },
      }
    }
  />
</div>
`;

exports[`ModalErrors Renders requestError correctly 1`] = `
<div
  className="alert-box alert margin-top"
>
  The form returned some errors
  <ul
    className="bullet"
  />
  <RequestError
    contact="cont@ct"
    error={
      Object {
        "response": Object {
          "data": Object {
            "issues": Array [
              Object {
                "field": null,
                "message": "Something happened in the back-end",
                "reference": null,
              },
            ],
          },
        },
      }
    }
  />
</div>
`;

exports[`ModalErrors Renders validationErrors correctly 1`] = `
<div
  className="alert-box alert margin-top"
>
  The form returned some errors
  <ul
    className="bullet"
  >
    <li
      key="0"
    >
      Validation failed.
    </li>
    <li
      key="1"
    >
      Another validation failed
    </li>
  </ul>
  <RequestError />
</div>
`;

exports[`Rendering the component A modal is in the close state by default 1`] = `
<Modal
  title="test"
>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="react-modal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={false}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "maxHeight": "688px",
          "width": "450px",
        },
      }
    }
    title="test"
  >
    <Portal
      containerInfo={
        <div
          class="react-modal"
        />
      }
    >
      <ModalPortal
        ariaHideApp={true}
        bodyOpenClassName="ReactModal__Body--open"
        closeTimeoutMS={0}
        defaultStyles={Object {}}
        isOpen={false}
        onRequestClose={[Function]}
        parentSelector={[Function]}
        portalClassName="react-modal"
        role="dialog"
        shouldCloseOnEsc={true}
        shouldCloseOnOverlayClick={false}
        shouldFocusAfterRender={true}
        shouldReturnFocusAfterClose={true}
        style={
          Object {
            "content": Object {
              "maxHeight": "688px",
              "width": "450px",
            },
          }
        }
        title="test"
      />
    </Portal>
  </Modal>
</Modal>
`;

exports[`Rendering the component A modal with a fixed width and height 1`] = `
<Modal
  height="350px"
  open={true}
  title="test"
  width="800px"
>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    isOpen={true}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="react-modal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={false}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "height": "350px",
          "maxHeight": "688px",
          "width": "800px",
        },
      }
    }
    title="test"
  >
    <Portal
      containerInfo={
        <div
          class="react-modal"
        >
          <div
            class="ReactModal__Overlay ReactModal__Overlay--after-open"
          >
            <div
              class="ReactModal__Content ReactModal__Content--after-open"
              role="dialog"
              style="width: 800px; max-height: 688px; height: 350px;"
              tabindex="-1"
            >
              <div
                class="modal-titlebar"
              >
                <div
                  class="modal-title"
                >
                  test
                </div>
                <div
                  class="modal-close-icon"
                >
                  <i
                    class="fa fa-times"
                  />
                </div>
              </div>
              <div
                class="modal-content"
              >
                <p>
                  Test
                </p>
              </div>
              <div
                class="modal-buttons"
              >
                <a
                  class="button modal-ok "
                  data-cy="modal-ok-button"
                >
                  OK
                  <div
                    class="bar"
                  />
                </a>
                <a
                  class="button secondary modal-cancel"
                >
                  Cancel
                </a>
              </div>
            </div>
          </div>
        </div>
      }
    >
      <ModalPortal
        ariaHideApp={true}
        bodyOpenClassName="ReactModal__Body--open"
        closeTimeoutMS={0}
        defaultStyles={Object {}}
        isOpen={true}
        onRequestClose={[Function]}
        parentSelector={[Function]}
        portalClassName="react-modal"
        role="dialog"
        shouldCloseOnEsc={true}
        shouldCloseOnOverlayClick={false}
        shouldFocusAfterRender={true}
        shouldReturnFocusAfterClose={true}
        style={
          Object {
            "content": Object {
              "height": "350px",
              "maxHeight": "688px",
              "width": "800px",
            },
          }
        }
        title="test"
      >
        <div
          className="ReactModal__Overlay ReactModal__Overlay--after-open"
          onClick={[Function]}
          onMouseDown={[Function]}
          style={Object {}}
        >
          <div
            className="ReactModal__Content ReactModal__Content--after-open"
            onClick={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            role="dialog"
            style={
              Object {
                "height": "350px",
                "maxHeight": "688px",
                "width": "800px",
              }
            }
            tabIndex="-1"
          >
            <div
              className="modal-titlebar"
            >
              <div
                className="modal-title"
              >
                test
              </div>
              <div
                className="modal-close-icon"
                onClick={[Function]}
              >
                <i
                  className="fa fa-times"
                />
              </div>
            </div>
            <div
              className="modal-content"
            >
              <p>
                Test
              </p>
            </div>
            <div
              className="modal-buttons"
            >
              <a
                className="button modal-ok "
                data-cy="modal-ok-button"
                onClick={[Function]}
              >
                OK
                <div
                  className="bar"
                />
              </a>
              <a
                className="button secondary modal-cancel"
                onClick={[Function]}
              >
                Cancel
              </a>
            </div>
          </div>
        </div>
      </ModalPortal>
    </Portal>
  </Modal>
</Modal>
`;

exports[`Rendering the component Rendering the closed state 1`] = `
<Modal
  open={false}
  title="test"
>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="react-modal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={false}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "maxHeight": "688px",
          "width": "450px",
        },
      }
    }
    title="test"
  >
    <Portal
      containerInfo={
        <div
          class="react-modal"
        />
      }
    >
      <ModalPortal
        ariaHideApp={true}
        bodyOpenClassName="ReactModal__Body--open"
        closeTimeoutMS={0}
        defaultStyles={Object {}}
        isOpen={false}
        onRequestClose={[Function]}
        parentSelector={[Function]}
        portalClassName="react-modal"
        role="dialog"
        shouldCloseOnEsc={true}
        shouldCloseOnOverlayClick={false}
        shouldFocusAfterRender={true}
        shouldReturnFocusAfterClose={true}
        style={
          Object {
            "content": Object {
              "maxHeight": "688px",
              "width": "450px",
            },
          }
        }
        title="test"
      />
    </Portal>
  </Modal>
</Modal>
`;

exports[`Rendering the component Rendering the open state 1`] = `
<Modal
  onClose={[Function]}
  open={true}
  title="test"
>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    isOpen={true}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="react-modal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={false}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "maxHeight": "688px",
          "width": "450px",
        },
      }
    }
    title="test"
  >
    <Portal
      containerInfo={
        <div
          class="react-modal"
        >
          <div
            class="ReactModal__Overlay ReactModal__Overlay--after-open"
          >
            <div
              class="ReactModal__Content ReactModal__Content--after-open"
              role="dialog"
              style="width: 450px; max-height: 688px;"
              tabindex="-1"
            >
              <div
                class="modal-titlebar"
              >
                <div
                  class="modal-title"
                >
                  test
                </div>
                <div
                  class="modal-close-icon"
                >
                  <i
                    class="fa fa-times"
                  />
                </div>
              </div>
              <div
                class="modal-content"
              >
                <p>
                  Test
                </p>
              </div>
              <div
                class="modal-buttons"
              >
                <a
                  class="button modal-ok "
                  data-cy="modal-ok-button"
                >
                  OK
                  <div
                    class="bar"
                  />
                </a>
                <a
                  class="button secondary modal-cancel"
                >
                  Cancel
                </a>
              </div>
            </div>
          </div>
        </div>
      }
    >
      <ModalPortal
        ariaHideApp={true}
        bodyOpenClassName="ReactModal__Body--open"
        closeTimeoutMS={0}
        defaultStyles={Object {}}
        isOpen={true}
        onRequestClose={[Function]}
        parentSelector={[Function]}
        portalClassName="react-modal"
        role="dialog"
        shouldCloseOnEsc={true}
        shouldCloseOnOverlayClick={false}
        shouldFocusAfterRender={true}
        shouldReturnFocusAfterClose={true}
        style={
          Object {
            "content": Object {
              "maxHeight": "688px",
              "width": "450px",
            },
          }
        }
        title="test"
      >
        <div
          className="ReactModal__Overlay ReactModal__Overlay--after-open"
          onClick={[Function]}
          onMouseDown={[Function]}
          style={Object {}}
        >
          <div
            className="ReactModal__Content ReactModal__Content--after-open"
            onClick={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            role="dialog"
            style={
              Object {
                "maxHeight": "688px",
                "width": "450px",
              }
            }
            tabIndex="-1"
          >
            <div
              className="modal-titlebar"
            >
              <div
                className="modal-title"
              >
                test
              </div>
              <div
                className="modal-close-icon"
                onClick={[Function]}
              >
                <i
                  className="fa fa-times"
                />
              </div>
            </div>
            <div
              className="modal-content"
            >
              <p>
                Test
              </p>
            </div>
            <div
              className="modal-buttons"
            >
              <a
                className="button modal-ok "
                data-cy="modal-ok-button"
                onClick={[Function]}
              >
                OK
                <div
                  className="bar"
                />
              </a>
              <a
                className="button secondary modal-cancel"
                onClick={[Function]}
              >
                Cancel
              </a>
            </div>
          </div>
        </div>
      </ModalPortal>
    </Portal>
  </Modal>
</Modal>
`;

exports[`Rendering the component Rendering the open state with an action in Progress 1`] = `
<Modal
  inProgress={true}
  onClose={[Function]}
  open={true}
  title="test"
>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    isOpen={true}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="react-modal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={false}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "maxHeight": "688px",
          "width": "450px",
        },
      }
    }
    title="test"
  >
    <Portal
      containerInfo={
        <div
          class="react-modal"
        >
          <div
            class="ReactModal__Overlay ReactModal__Overlay--after-open"
          >
            <div
              class="ReactModal__Content ReactModal__Content--after-open"
              role="dialog"
              style="width: 450px; max-height: 688px;"
              tabindex="-1"
            >
              <div
                class="modal-titlebar"
              >
                <div
                  class="modal-title"
                >
                  test
                </div>
                <div
                  class="modal-close-icon"
                >
                  <i
                    class="fa fa-times"
                  />
                </div>
              </div>
              <div
                class="modal-content"
              >
                <p>
                  Test
                </p>
              </div>
              <div
                class="modal-buttons"
              >
                <a
                  class="button modal-ok in-progress"
                  data-cy="modal-ok-button"
                >
                  OK
                  <div
                    class="bar"
                  />
                </a>
                <a
                  class="button secondary modal-cancel"
                >
                  Cancel
                </a>
              </div>
            </div>
          </div>
        </div>
      }
    >
      <ModalPortal
        ariaHideApp={true}
        bodyOpenClassName="ReactModal__Body--open"
        closeTimeoutMS={0}
        defaultStyles={Object {}}
        isOpen={true}
        onRequestClose={[Function]}
        parentSelector={[Function]}
        portalClassName="react-modal"
        role="dialog"
        shouldCloseOnEsc={true}
        shouldCloseOnOverlayClick={false}
        shouldFocusAfterRender={true}
        shouldReturnFocusAfterClose={true}
        style={
          Object {
            "content": Object {
              "maxHeight": "688px",
              "width": "450px",
            },
          }
        }
        title="test"
      >
        <div
          className="ReactModal__Overlay ReactModal__Overlay--after-open"
          onClick={[Function]}
          onMouseDown={[Function]}
          style={Object {}}
        >
          <div
            className="ReactModal__Content ReactModal__Content--after-open"
            onClick={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            role="dialog"
            style={
              Object {
                "maxHeight": "688px",
                "width": "450px",
              }
            }
            tabIndex="-1"
          >
            <div
              className="modal-titlebar"
            >
              <div
                className="modal-title"
              >
                test
              </div>
              <div
                className="modal-close-icon"
                onClick={[Function]}
              >
                <i
                  className="fa fa-times"
                />
              </div>
            </div>
            <div
              className="modal-content"
            >
              <p>
                Test
              </p>
            </div>
            <div
              className="modal-buttons"
            >
              <a
                className="button modal-ok in-progress"
                data-cy="modal-ok-button"
                onClick={null}
              >
                OK
                <div
                  className="bar"
                />
              </a>
              <a
                className="button secondary modal-cancel"
                onClick={[Function]}
              >
                Cancel
              </a>
            </div>
          </div>
        </div>
      </ModalPortal>
    </Portal>
  </Modal>
</Modal>
`;
